### ISSUE TITLE
Separate measured savings from modeled tokens avoided

### ISSUE THREAD

**Issue by styler-ai** [styler-ai/ProjectAtlas#246]
## Summary

Introduce layered token accounting for ProjectAtlas reports, using Headroom as the reference example for separating measured savings from modeled/counterfactual savings. ProjectAtlas should still include `tokens_avoided`, but it should be reported as an explicitly estimated metric with a clear denominator, confidence level, and dedupe scope.

Related: #245 covers calibrated/deduped token telemetry generally. This issue focuses on the reporting architecture and semantics for `tokens_avoided`.

## Goal

Make `projectatlas token` credible when savings look very high by separating:

- `measured_tokens_saved`: before/after token deltas where both sides are observable.
- `modeled_tokens_avoided`: estimated context the agent likely did not need to emit/read because ProjectAtlas narrowed the workflow.
- `deduped_tokens_avoided`: session/task-scoped avoided tokens after removing repeated claims over the same baseline.
- `gross_tokens_avoided`: cumulative/non-deduped estimate retained for diagnostics, not the headline number.

Headroom is a useful example because it keeps active compression savings and counterfactual/output savings conceptually separate, uses explicit before/after denominators for measured compression, and labels modeled savings as modeled rather than exact.

## Non-goals

- Do not remove `tokens_avoided` from reports.
- Do not claim billing-token accuracy unless an actual model tokenizer is used.
- Do not make network calls or depend on external telemetry services.
- Do not turn Headroom internals into a ProjectAtlas dependency.

## Requirements

- Keep `tokens_avoided` in CLI/TOON/JSON token reports.
- Split reported totals by accounting layer:
  - `observed_delta`
  - `modeled_avoidance`
  - `deduped_modeled_avoidance`
  - `gross_modeled_avoidance`
- For every token event, persist enough metadata to explain the number:
  - `estimate_method`
  - `denominator_kind`
  - `baseline_identity`
  - `baseline_fingerprint`
  - `dedupe_scope`
  - `confidence`
  - `tokenizer_backend`
- Headline savings should prefer conservative/deduped numbers. Gross lifetime estimates can remain visible as secondary diagnostics.
- Negative deltas should be represented honestly instead of clamped into savings.
- Current heuristic estimates must remain labeled as estimates, not exact token counts.

## Constraints / Assumptions

- Current ProjectAtlas token estimates are intentionally offline and heuristic-heavy, commonly `ceil(chars / 4)` or `ceil(bytes / 4)` depending on the baseline.
- Some ProjectAtlas value is real avoidance: the agent may inspect an overview, folder ranking, summary, outline, or slice instead of reading many full files.
- That value is counterfactual. It needs a modeled denominator and dedupe rules, otherwise repeated operations can inflate cumulative savings.
- The architecture should be Rust-native and fit the existing `projectatlas-core`, `projectatlas-db`, and `projectatlas-cli` telemetry boundaries.

## UX / Design References

`projectatlas token` should make the headline hard to misread. Example shape:

```toon
token_report:
  headline:
    measured_tokens_saved: 12400
    deduped_tokens_avoided_estimate: 38200
    estimate_basis: heuristic_estimate
    confidence: mixed
  diagnostics:
    gross_tokens_avoided_estimate: 156000
    repeated_baselines_deduped: 9
    tokenizer_backend: heuristic_chars_per_token
```

## Data / API Notes

Suggested event model additions:

```rust
enum TokenAccountingLayer {
    ObservedDelta,
    ModeledAvoidance,
    DedupedModeledAvoidance,
    GrossModeledAvoidance,
}

enum TokenConfidence {
    Observed,
    CalibratedEstimate,
    HeuristicEstimate,
    PolicyEstimate,
}

enum TokenDenominatorKind {
    FullFile,
    CandidateSet,
    DirectoryWalk,
    SearchSpace,
    AgentMessage,
}
```

Each event should carry a stable baseline key, for example:

```text
baseline_identity = "navigation_avoidance:query=token telemetry:folder=crates/projectatlas-core"
baseline_fingerprint = hash(sorted(candidate_file_ids + mtimes_or_content_hashes))
dedupe_scope = "session"
```

## Implementation Plan

1. Add accounting-layer and confidence metadata to token telemetry structs and SQLite persistence.
2. Classify existing events:
   - summary/outline/slice vs full-file baseline -> `ObservedDelta` or high-confidence estimate when the full file size is known.
   - folder/files/search narrowing vs candidate set -> `ModeledAvoidance`.
   - repeated same baseline in one session/task -> also contribute to gross, but only once to deduped totals.
3. Update aggregation so report totals expose measured, modeled, deduped, and gross numbers separately.
4. Update CLI/JSON/TOON output to make deduped estimates the headline and gross estimates secondary.
5. Add docs explaining why high gross savings can be valid but should not be read as exact billing-token savings.

## Code Touch Points

- `crates/projectatlas-core/src/telemetry.rs`
- `crates/projectatlas-cli/src/runtime.rs`
- `crates/projectatlas-db/src/lib.rs`
- `docs/benchmarks/` or equivalent telemetry documentation
- CLI/e2e tests that snapshot `projectatlas token` output

## Tests & Verification

- Unit test measured full-file compression where both baseline and emitted text are known.
- Unit test modeled navigation avoidance where the candidate set is known but counterfactual.
- Unit test repeated identical baseline events increase gross totals but not deduped session totals.
- Unit test modified baseline fingerprint allows a new deduped estimate.
- Unit test negative deltas are reported as expansion/negative savings.
- CLI test verifies `projectatlas token` labels heuristic estimates clearly.

## Rollout / Monitoring

- Preserve old fields for one release if needed, but mark them as gross/legacy estimates.
- Add a short migration note for interpreting old telemetry numbers.
- Keep the new report schema backward-compatible where practical.

## Risks / Edge Cases

- Dedupe can undercount legitimate repeated value if scoped too broadly.
- Gross totals can still look high, so the UI must avoid making them the main success metric.
- Tokenizer calibration can change historical comparisons; reports need to include the tokenizer/estimate method.
- Some workflows have no reliable counterfactual baseline and should be reported as policy estimates only.

## Acceptance Criteria

- `projectatlas token` still includes `tokens_avoided`.
- `tokens_avoided` is clearly labeled as estimated/modeled when it is not directly measured.
- Reports separate measured saved tokens from modeled avoided tokens.
- Reports expose both deduped and gross avoided-token estimates.
- Repeated same-baseline operations do not inflate the deduped headline.
- Tests cover measured compression, modeled avoidance, dedupe, baseline changes, and negative deltas.
- Documentation explains the Headroom-inspired accounting distinction without depending on Headroom.

**Comment by styler-ai** [comment]
Implementation note for the terminal dashboard:

Use `ratatui` for the `projectatlas token --view tui` renderer, with `crossterm` as the terminal backend/event layer. This keeps the feature cross-platform for PowerShell, Windows Terminal, Git Bash, WSL, Linux, and macOS terminals without separate shell-specific implementations.

Preferred shape:

```text
projectatlas token
  -> existing structured agent/JSON/TOON output

projectatlas token --view tui
  -> ratatui dashboard
     -> summary gauge
     -> measured vs modeled bars
     -> deduped tokens avoided bar chart
     -> gross avoided estimate as secondary diagnostic
     -> trend sparkline
     -> bucket table
```

Keep the calculation in the existing telemetry/reporting layer. The TUI should only render `TokenOverview` / `TokenTrendReport`; it should not calculate savings itself.

Useful widgets:

- `BarChart` for measured/modelled/avoided-token comparisons.
- `Chart` for trend windows.
- `Sparkline` for compact history.
- `Gauge` / `LineGauge` for savings percentages.
- `Table` for bucket breakdowns.

Do not hand-roll chart rendering unless `ratatui` cannot cover a specific display need. The goal is a genuinely readable terminal dashboard, not a prettier version of the current text output.

**Comment by styler-ai** [comment]
Additional dashboard quality bar:

The TUI should be designed as a terminal-native dashboard, not a fake web UI. `ratatui` can look good here if the first screen is restrained and chart-focused.

Target layout:

```text
ProjectAtlas Token Report
Session: current                         Basis: heuristic + modeled estimates

Saved vs Avoided
Measured saved          ████████████░░░░░░░░   12.4k
Deduped avoided est.    ██████████████████░░   38.2k
Gross avoided est.      █████████████████████  156.0k

Savings Rate
Measured     42%  █████████░░░░░░░░░░░
Modeled      71%  ███████████████░░░░░

Trend
▁▂▃▄▆▇▅▆█▇▆▄▅▇█

Buckets
┌──────────────────────┬──────────┬────────────┬────────────┐
│ Bucket               │ Before   │ After      │ Saved      │
├──────────────────────┼──────────┼────────────┼────────────┤
│ file summaries       │ 48.1k    │ 9.8k       │ 38.3k      │
│ slices               │ 21.0k    │ 4.2k       │ 16.8k      │
│ search narrowing     │ modeled  │ modeled    │ 31.4k est. │
└──────────────────────┴──────────┴────────────┴────────────┘
```

Design rules:

- One useful screen first; avoid nested menus unless a real need appears.
- Big bars for headline numbers.
- Compact table for bucket detail.
- Clear labels: `measured`, `deduped estimate`, `gross estimate`.
- Light color use only: green for measured savings, cyan/blue for modeled avoided, gray/yellow for estimate warnings.
- Unicode charts by default with ASCII fallback for limited terminals.
- No fake precision: prefer `38.2k est.` over exact-looking modeled numbers.

Acceptance criterion: the TUI must make token telemetry easier to trust, not merely prettier.

**Comment by styler-ai** [comment]
Implemented in the current ProjectAtlas working tree and verified locally.

Coverage included:
- cargo fmt --all -- --check
- cargo check --workspace --all-targets --all-features
- cargo clippy --workspace --all-targets --all-features -- -D warnings
- cargo test --workspace --all-features
- cargo test --doc --all-features
- RUSTDOCFLAGS=-D warnings cargo doc --workspace --no-deps --all-features
- cargo deny check
- projectatlas scan + lint --report-untracked --purpose-level low
- live token TUI overview, trend, and tokenizer-calibrated render checks

Closed as completed per owner request.
